File: trunk/NowPlaying/Android/src/org/metasyntactic/caches/TrailerCache.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

package org.metasyntactic.caches;

import org.metasyntactic.Application;
import org.metasyntactic.Constants;
import org.metasyntactic.collections.BoundedPrioritySet;
import org.metasyntactic.data.Movie;
import org.metasyntactic.threading.ThreadingUtilities;
import org.metasyntactic.utilities.FileUtilities;
import org.metasyntactic.utilities.NetworkUtilities;
import org.metasyntactic.utilities.difference.EditDistance;

import java.io.File;
import java.util.*;

public class TrailerCache extends AbstractCache {
  private final BoundedPrioritySet<Movie> prioritizedMovies = new BoundedPrioritySet<Movie>(9);

  private String trailerFileName(final Movie movie) {
    return FileUtilities.sanitizeFileName(movie.getCanonicalTitle());
  }

  private File trailerFilePath(final Movie movie) {
    return new File(Application.trailersDirectory, trailerFileName(movie));
  }

  protected void clearStaleDataBackgroundEntryPoint() {
    clearDirectory(Application.trailersDirectory);
  }

  @SuppressWarnings("unchecked")
  private List<List<Movie>> getOrderedMovies(final List<Movie> movies) {
    final List<Movie> moviesWithoutTrailers = new ArrayList<Movie>();
    final List<Movie> moviesWithTrailers = new ArrayList<Movie>();

    final long now = new Date().getTime();

    for (final Movie movie : movies) {
      final File file = trailerFilePath(movie);
      if (!file.exists()) {
        moviesWithoutTrailers.add(movie);
      } else {
        final long writeTime = file.lastModified();
        final long span = Math.abs(writeTime - now);

        if (span > 2 * Constants.ONE_DAY) {
          moviesWithTrailers.add(movie);
        }
      }
    }

    return Arrays.asList(moviesWithoutTrailers, moviesWithTrailers);
  }

  public void update(final List<Movie> movies) {
    final Runnable runnable = new Runnable() {
      public void run() {
        updateBackgroundEntryPoint(movies);
      }
    };
    ThreadingUtilities.performOnBackgroundThread("Update Trailers", runnable, this.lock, false/*visible*/);
  }

  private void updateBackgroundEntryPoint(final List<Movie> movies) {
    final List<List<Movie>> orderedMovies = getOrderedMovies(movies);

    final String url = "http://" + Application.host + ".appspot.com/LookupTrailerListings?q=index";
    final String indexText = NetworkUtilities.downloadString(url, false);
    if (indexText == null) {
      return;
    }

    final Map<String, List<String>> index = generateIndex(indexText);

    for (final List<Movie> values : orderedMovies) {
      downloadTrailers(values, index);
    }
  }

  private void downloadMovieTrailer(final Movie movie, final Map<String, List<String>> index) {
    if (movie == null) {
      return;
    }

    final String bestKey = EditDistance.findClosestMatch(movie.getCanonicalTitle().toLowerCase(), index.keySet());
    if (bestKey == null) {
      // no trailer for this movie.  record that fact.  we'll try again later
      FileUtilities.writeStringCollection(new ArrayList<String>(), trailerFilePath(movie));
      return;
    }

    final List<String> studioAndLocation = index.get(bestKey);
    final String studio = studioAndLocation.get(0);
    final String location = studioAndLocation.get(1);

    final String url = "http://" + Application
        .host + ".appspot.com/LookupTrailerListings?studio=" + studio + "&name=" + location;
    final String trailersString = NetworkUtilities.downloadString(url, false);

    if (trailersString == null) {
      // didn't get any data.  ignore this for now.
      return;
    }

    final List<String> trailers = Arrays.asList(trailersString.split("\n"));
    FileUtilities.writeString(trailers.get(0), trailerFilePath(movie));
    Application.refresh();
  }

  private void downloadTrailers(final List<Movie> movies, final Map<String, List<String>> index) {
    final Set<Movie> moviesSet = new TreeSet<Movie>(movies);

    Movie movie;
    do {
      movie = this.prioritizedMovies.removeAny(moviesSet);
      downloadMovieTrailer(movie, index);
    } while (movie != null && !this.shutdown);
  }

  private Map<String, List<String>> generateIndex(final String indexText) {
    final Map<String, List<String>> index = new HashMap<String, List<String>>();

    for (final String row : indexText.split("\n")) {
      final String[] values = row.split("\t");
      if (values.length != 3) {
        continue;
      }

      final String fullTitle = values[0];
      final String studio = values[1];
      final String location = values[2];

      index.put(fullTitle.toLowerCase(), Arrays.asList(studio, location));
    }

    return index;
  }

  public String getTrailer(final Movie movie) {
    return FileUtilities.readString(trailerFilePath(movie));
  }

  public void prioritizeMovie(final Movie movie) {
    if (trailerFilePath(movie).exists()) {
      return;
    }

    this.prioritizedMovies.add(movie);
  }
}

File: NowPlaying/iPhone/Classes/TrailerCache.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "TrailerCache.h"

#import "Application.h"
#import "DifferenceEngine.h"
#import "FileUtilities.h"
#import "LinkedSet.h"
#import "Movie.h"
#import "NetworkUtilities.h"
#import "NowPlayingAppDelegate.h"
#import "ThreadingUtilities.h"

@implementation TrailerCache

@synthesize model;
@synthesize gate;
@synthesize prioritizedMovies;

- (void) dealloc {
    self.model = nil;
    self.gate = nil;
    self.prioritizedMovies = nil;

    [super dealloc];
}


- (id) initWithModel:(NowPlayingModel*) model_ {
    if (self = [super init]) {
        self.model = model_;
        self.gate = [[[NSRecursiveLock alloc] init] autorelease];
        self.prioritizedMovies = [LinkedSet setWithCountLimit:8];
    }

    return self;
}


+ (TrailerCache*) cacheWithModel:(NowPlayingModel*) model {
    return [[[TrailerCache alloc] initWithModel:model] autorelease];
}


- (NSString*) trailerFile:(Movie*) movie {
    NSString* name = [[FileUtilities sanitizeFileName:movie.canonicalTitle] stringByAppendingPathExtension:@"plist"];
    return [[Application trailersDirectory] stringByAppendingPathComponent:name];
}


- (void) deleteObsoleteTrailers:(NSArray*) movies {
    NSArray* paths = [FileUtilities directoryContentsPaths:[Application trailersDirectory]];
    NSMutableSet* set = [NSMutableSet setWithArray:paths];

    for (Movie* movie in movies) {
        NSString* filePath = [self trailerFile:movie];
        [set removeObject:filePath];
    }

    for (NSString* filePath in set) {
        NSDate* downloadDate = [FileUtilities modificationDate:filePath];

        if (downloadDate != nil) {
            if (ABS(downloadDate.timeIntervalSinceNow) > ONE_MONTH) {
                [FileUtilities removeItem:filePath];
            }
        }
    }
}


- (NSArray*) getOrderedMovies:(NSArray*) movies {
    NSMutableArray* moviesWithoutTrailers = [NSMutableArray array];
    NSMutableArray* moviesWithTrailers = [NSMutableArray array];

    for (Movie* movie in movies) {
        NSDate* downloadDate = [FileUtilities modificationDate:[self trailerFile:movie]];

        if (downloadDate == nil) {
            [moviesWithoutTrailers addObject:movie];
        } else {
            NSTimeInterval span = downloadDate.timeIntervalSinceNow;
            if (ABS(span) > (3 * ONE_DAY)) {
                [moviesWithTrailers addObject:movie];
            }
        }
    }

    return [NSArray arrayWithObjects:moviesWithoutTrailers, moviesWithTrailers, nil];
}


- (void) update:(NSArray*) movies {
    [ThreadingUtilities performSelector:@selector(backgroundEntryPoint:)
                               onTarget:self
               inBackgroundWithArgument:movies
                                   gate:gate
                                visible:NO];
}


- (void) downloadMovieTrailer:(Movie*) movie
                        index:(NSDictionary*) index
                    indexKeys:(NSArray*) indexKeys
                       engine:(DifferenceEngine*) engine {
    if (movie == nil) {
        return;
    }

    NSInteger arrayIndex = [engine findClosestMatchIndex:movie.canonicalTitle.lowercaseString inArray:indexKeys];
    if (arrayIndex == NSNotFound) {
        // no trailer for this movie.  record that fact.  we'll try again later
        [FileUtilities writeObject:[NSArray array] toFile:[self trailerFile:movie]];
        return;
    }

    NSArray* studioAndLocation = [index objectForKey:[indexKeys objectAtIndex:arrayIndex]];
    NSString* studio = [studioAndLocation objectAtIndex:0];
    NSString* location = [studioAndLocation objectAtIndex:1];

    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupTrailerListings?studio=%@&name=%@", [Application host], studio, location];
    NSString* trailersString = [NetworkUtilities stringWithContentsOfAddress:url
                                                                   important:NO];
    if (trailersString == nil) {
        // didn't get any data.  ignore this for now.
        NSLog(@"", nil);
        return;
    }

    NSArray* trailers = [trailersString componentsSeparatedByString:@"\n"];
    [FileUtilities writeObject:trailers toFile:[self trailerFile:movie]];
    [NowPlayingAppDelegate refresh];
}


- (Movie*) getNextMovie:(NSMutableArray*) movies {
    Movie* movie = [prioritizedMovies removeLastObjectAdded];

    if (movie != nil) {
        return movie;
    }

    if (movies.count > 0) {
        movie = [[[movies lastObject] retain] autorelease];
        [movies removeLastObject];
        return movie;
    }

    return nil;
}


- (void) downloadTrailers:(NSMutableArray*) movies
                    index:(NSDictionary*) index
                indexKeys:(NSArray*) indexKeys {
    DifferenceEngine* engine = [DifferenceEngine engine];

    Movie* movie;
    do {
        NSAutoreleasePool* autoreleasePool= [[NSAutoreleasePool alloc] init];
        {
            movie = [self getNextMovie:movies];
            [self downloadMovieTrailer:movie
                                 index:index
                             indexKeys:indexKeys
                                engine:engine];
        }
        [autoreleasePool release];
    } while (movie != nil);
}


- (void) prioritizeMovie:(Movie*) movie {
    [prioritizedMovies addObject:movie];
}


- (NSDictionary*) generateIndex:(NSString*) indexText {
    NSMutableDictionary* index = [NSMutableDictionary dictionary];

    NSArray* rows = [indexText componentsSeparatedByString:@"\n"];
    for (NSString* row in rows) {
        NSArray* values = [row componentsSeparatedByString:@"\t"];
        if (values.count != 3) {
            continue;
        }

        NSString* fullTitle = [values objectAtIndex:0];
        NSString* studio = [values objectAtIndex:1];
        NSString* location = [values objectAtIndex:2];

        [index setObject:[NSArray arrayWithObjects:studio, location, nil]
                  forKey:fullTitle.lowercaseString];
    }

    return index;
}


- (void) backgroundEntryPoint:(NSArray*) movies {
    [self deleteObsoleteTrailers:movies];

    NSArray* orderedMovies = [self getOrderedMovies:movies];
    NSMutableArray* moviesWithoutTrailers = [orderedMovies objectAtIndex:0];
    NSMutableArray* moviesWithTrailers = [orderedMovies objectAtIndex:1];
    if (moviesWithoutTrailers.count == 0 && moviesWithTrailers.count == 0) {
        return;
    }

    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupTrailerListings?q=index", [Application host]];
    NSString* indexText = [NetworkUtilities stringWithContentsOfAddress:url important:NO];
    if (indexText == nil) {
        return;
    }

    NSDictionary* index = [self generateIndex:indexText];
    NSArray* indexKeys = index.allKeys;

    [self downloadTrailers:moviesWithoutTrailers index:index indexKeys:indexKeys];
    [self downloadTrailers:moviesWithTrailers index:index indexKeys:indexKeys];
}


- (NSArray*) trailersForMovie:(Movie*) movie {
    NSArray* trailers = [FileUtilities readObject:[self trailerFile:movie]];
    if (trailers == nil) {
        return [NSArray array];
    }
    return trailers;
}


@end